README.md

```diff
@@ -33,28 +49,47 @@ Think of __Stroke__ as the amplitude and __Depth__ a linear offset that is added
+One design goal was to have a unobtrusive failure handling when invalid parameters are given. Either from the user with values that lay outside the physics of the machine, or from a pattern commanding an impossible speed, position or acceleration. 
 
-Also on the pattern side `constrain()` is used to ensure no impossible motion commands leading to crashes or step losses are executed. This manifests in a distortion of the motion. Strokes may be shortened when position targets outside of the machine bounds are requested (e.g. `stroke > depth`). Acceleration and speed are limited leading to  distorted ramps. The motion is executed over the full distance, but may take slightly longer then expected to reach the target position. 
+All `MotorInterface` implementations are expected to constrain any Motion Command which will exit the provided `motionBounds`. The command will still execute, but the full range of motion will be cut short. This manifests in a distortion of the motion. Strokes may be shortened when position targets outside of the machine bounds are requested (e.g. `stroke > depth`). Acceleration and speed are limited leading to distorted ramps. The motion is executed over the full distance, but may take slightly longer then expected to reach the target position. 
 
 ### Mid-Stroke Parameter Update
 It is possible to update any parameter like depth, stroke, speed and pattern mid-stroke. This gives a very responsive and fluid user experience. Safeguards are in place to ensure the move stays inside the bounds of the machine at any time.
 
-### State Machine
-An internal finite state machine handles the different states of the machine. See the below graph with all functions relating to the state machine and how to cause transitions:
-
-![State Machine](./doc/state-machine.svg)
-* __UNDEFINED:__ The initial state prior to homing. Stepper / Servo are disabled and the position is undefined.
-* __READY:__ Homing defines the position inside the internal coordinate system. Machine is now ready to be used and accepts motion commands.
-* __PATTERN:__ The cyclic motion has started and the pattern generator is commanding a sequence of trapezoidal motions until stopped.
-* __SETUPDEPTH:__ The servo always follows the depth position. This can be used to setup the optimal stroke depth. 
-
 ## Usage
 StrokeEngine aims to have a simple and straight forward, yet powerful API. The following describes the minimum case to get up and running. All input parameters need to be specified in real world (metric) units.
-### Initialize
+
+### > Initialize
 First all parameters of the machine and the servo need to be set. Including the pins for interacting with the driver and an (optionally) homing switch.
 ```cpp
-#include <StrokeEngine.h>
+#include "motor_stepper.hpp"
```

Comment:
I would highly prefer, if this sample code could only contain lines which are relevant for the library. I find it easier for the novice if it does not contain unneeded bloat (a.k.a. blynk, wifi, etc.)

Reply:
Yeah, that's fair enough. Especially in a README I should probably be concise and clear, rather than trying to cover a full example. 

I'll move the blynk/wifi in a `examples/blynk.cpp` file somewhere, as I do think it can provide a really easy way to be up and running with a mobile app controller.

Reply:
And the motor-folder could come one level up with all supported motors. I'll definitively write an own driver for the JMC iHSV57 v6 (should you buy one make sure you get firmware v6, as they changed the protocol from v5)

Reply:
I've adjusted the README to use only a simple example of interfacing with the StrokeEngine.
Currently, I've been building my LinMot driver in CANFuck, and I think that would be a good place for the iHSV57 driver as well.

That allows StrokeEngine to remain decoupled from any of the devices. Otherwise I'd need to move a bunch of CANOpen initialization and setup into this repo.

Let me know if that's OK with you

Reply:
If you haven't checked out my companion repo that I've been using for StrokeEngine integration, you should.
https://github.com/zylos146/CANFuck

Reply:
Or maybe CANFuck should really just turn into a OSSM-Stroke v2 repo? It's really just an improvement on the OSSM codebase, and I hope to replicate the rest of the current OSSM funtionality sometime soon too (ESPNow controllers, etc)

Reply:
I was always my design goal to have StrokeEngine independent of any other code (except the libraries it's using). That's why there is no communication layer. Also the API is structured in a way that it comes with the batteries included. In that sense the explicit motor drivers should be an integral part of StrokeEngine. Otherwise it will be extremely hard for the typical Arduino coder to integrate everything to write his own code. StrokeEngine is really just a library. UI, control, communication should all be part of the main program.

I would appreciate it, if you could refactor your CANfuck in a way that the motor controllers can live inside StrokeEngine without any further dependency (except FastAccelStepper/modbus and your CAN/linmot llibraries) on the host code. A.k.a your last comment. CANFuck becomes one more viable firmware choice like OSSM, Ortlofs OSSM fork, FuckIO, ...

If the motor driver is not part of StrokeEngine this will break the code for many firmware's currently depending on StrokeEngine.

Reply:
Yeah, I should be able to do something. It'll introduce a dependency on CANOpenNode into StrokeEngine, but it should be minimal and optional. I've got the CANOpenNode mostly encapsulated in https://github.com/zylos146/CANOpenNode-ESP32 already, so it should hopefully not be too hard moving it over

Reply:
I think the biggest issue is gonna be figuring out how to handle library dependenies. CANOpenNode is a PlatformIO Library, as it gives control over build flags, and lets me include the Object Dictionary and other config from outside the lib directory.

CANOpenNode isn't really built to be encapsulated in a Library, as it needs includes for these files, so I'll have to see what I can do. I'll find something. Worst-case CANOpenNode is just a Git Submodule inside StrokeEngine, and is directly included into the src directory

Reply:
Yeah, I think the Git Submodule might be the best bet honestly. Keeps it simple

Reply:
actually, that wouldn't solve the issue, because the OD is in CANFuck, or whatever project uses StrokeEngine

Reply:
> I think the biggest issue is gonna be figuring out how to handle library dependenies. CANOpenNode is a PlatformIO Library, as it gives control over build flags, and lets me include the Object Dictionary and other config from outside the lib directory.
> 
> CANOpenNode isn't really built to be encapsulated in a Library, as it needs includes for these files, so I'll have to see what I can do. I'll find something. Worst-case CANOpenNode is just a Git Submodule inside StrokeEngine, and is directly included into the src directory

What does the code do that you would need to depend on? What is OD? 

StrokeEngine could also become a platform.io library, if that would easy anything.

Reply:
The Object Dictionary is a description of all the parameters of an Object in CANOpen. If I open the device.xdd file in CANFuck in the ODEditor provided by CANOpenNode, I can see 
![image](https://user-images.githubusercontent.com/63624195/172002298-8bac648e-6ec7-43d4-8bf8-3804af0ed887.png)

The Object Dictionary also configures RPDOs and TPDOs which handle automatic transmission of these parameters between different Nodes on the CANOpen network. I just specify what IDs we need to send the data to, which parameters to pack, and what triggers the event
![image](https://user-images.githubusercontent.com/63624195/172002572-d45ce2c1-f573-47aa-b261-163b6aefe590.png)

This is usually provided by the end-user, as they'll need to add any custom parameters and PDOs into it.
There isn't any real option of the OD being provided by a Library, then extended by the end-user.

Which definitely makes the CANOpen-based implementations more complex.
It also makes the includes and compilation trickier as well, as the CANOpenNode we attempt to isolate in a library, but it needs all of this include access to external headers defined in the root project

Reply:
PlatformIO libraries let me bypass the typical boundary of the library, by including a build flag of 
```"build": { "flags": [ "-I ../../include" ] }```

So making this a PlatformIO library would absolutely help.
I don't know of a good solution for getting this to compile correctly on Arduino, as it doesn't support PlatformIO style libraries as far as I know.

I'm gonna continue to poke around and think of solutions

Reply:
I might be able to provide a stripped down CANOpenNode-ESP32-Headers implementation which provides the important data types like `OD_extension_t` and functions like `OD_set_u16`

Reply:
I was trying to keep the CANOpenNode implementation very minimal on modifications though, so it could be easily updated to a newer version just by copy/pasting from the CANOpenNode root repository

Reply:
What kind of data needs to be defined in the user code? Do you use it for more then just talking to the linmot? If the motor is the only thing using CAN this should be doable, isn't it?

Reply:
So technically yeah, we could embed the OD into StrokeEngine without problem, but I might want to use CANOpen for talking to other hardware at some point.

What I've thought up is actually splitting the Drivers out from StrokeEngine itself, into separate libraries. You'd have StrokeEngine which is the core library with some generic Drivers. Then you'd have 
- StrokeEngine-CANOpen (LinMot, Future CANOpen Drives)
- StrokeEngine-iHSV57
- StrokeEngine-BLDC

That would also allow others to easily develop their own drivers and expose them for others to use, without needing to get a PR accepted into StrokeEngine.

For using with an Arduino, the code is still super simple. It's just

```
#include "StrokeEngine.h"
#include "StrokeEngine-iHSV57.h"

// Init Stroke engine using example code from README
```

That way I can keep the large CANOpen implementation out of StrokeEngine, which is really all I wanted in the first place.
the StrokeEngine-CANOpen library can be a PlatformIO-only library so I can take advantage of the build-flags for allowing user-specified OD.

Reply:
The rest of the Libraries will remain Arduino Libraries so both the Arduino IDE, and PlatformIO can use them

Reply:
I'm thinking of a future where this pattern can be extended to Motion Patterns as well.
So someone can develop their own `StrokeEngine-Throat` Pattern Library that provides a set of new patterns designed for throat-fucking, and they will be automatically registered with StrokeEngine Core so any UI/Controller can use them

Reply:
For reference, I was thinking of adding a Force Sensor into my Linmot-based fucking machine, and I was gonna interface that via CANOpen. So thats an example where I want to make sure the user can define their own OD

Reply:
Honestly, I do think defining their own OD is gonna be a minimal amount of users though. It's quite complex technically, and I wouldn't be surprised if I ended up being the only one interfacing CANOpen devices. I think a lot of others would rather just interface a Force Sensor via an Analog Pin, or something easier.

Reply:
> So technically yeah, we could embed the OD into StrokeEngine without problem, but I might want to use CANOpen for talking to other hardware at some point.
> 
> What I've thought up is actually splitting the Drivers out from StrokeEngine itself, into separate libraries. You'd have StrokeEngine which is the core library with some generic Drivers. Then you'd have 
> - StrokeEngine-CANOpen (LinMot, Future CANOpen Drives)
> - StrokeEngine-iHSV57
> - StrokeEngine-BLDC
> 
> That would also allow others to easily develop their own drivers and expose them for others to use, without needing to get a PR accepted into StrokeEngine.
> 
> For using with an Arduino, the code is still super simple. It's just
> 
> ```
> #include "StrokeEngine.h"
> #include "StrokeEngine-iHSV57.h"
> 
> // Init Stroke engine using example code from README
> ```
> 
> That way I can keep the large CANOpen implementation out of StrokeEngine, which is really all I wanted in the first place.
> the StrokeEngine-CANOpen library can be a PlatformIO-only library so I can take advantage of the build-flags for allowing user-specified OD.

I think this sounds like a real good compromise. Then you can do your CAN thing without creating unreached dependencies for StrokeEngine. Anyone with special driver needs can easily attach to StrokeEngine API. And I can ship with "batteries included" for the most popular drivers like generic stepper and ihsv57.

Reply:
> I'm thinking of a future where this pattern can be extended to Motion Patterns as well.
> So someone can develop their own `StrokeEngine-Throat` Pattern Library that provides a set of new patterns designed for throat-fucking, and they will be automatically registered with StrokeEngine Core so any UI/Controller can use them

That was actually my idea, too. I wanted to strip the basic patterns to a bare minimum that are fixed and the remaining pattern libraries, whether statically coded or scripted can be included by a similar mechanism.

Reply:
Also I would simplify homing to just one virtual function which must be overridden in the motor driver. Then it's up to the motor driver to determine how to home best.

Reply:
Alright, yeah that sounds like the best approach. I'll move the stubs for the generic/iHSV57 driver into StrokeEngine in this PR, and I'll split my CANOpen/LinMot support out to StrokeEngine-CANOpen.

I'll make `goToHome` a virtual function then. I'll let the individual drivers handle moving into HOMING state, and activating the HOMED flags.